readme.md
# Laravel OpenaiAttribute

Allows defining OpenAI generated attributes for Laravel model.

[![GitHub license](https://img.shields.io/github/license/ruvents/runet-id-php-client.svg?style=flat-square)](https://github.com/dmsemenov/openai-attribute/blob/master/license.md)
[![Latest Version on Packagist][ico-version]][link-packagist]
[![Total Downloads][ico-downloads]][link-downloads]

This is where your description should go. Take a look at [contributing.md](contributing.md) to see a to do list.

## Installation

- Install package
``` bash
 composer require dmsemenov/openai-attribute
```
- Add your OpenAI api key `OPENAI_API_KEY` to `.env`.    
Key can be found here https://platform.openai.com/account/api-keys

## Configuration
 
Publish config if needed:
``` bash
 php artisan vendor:publish --provider="Dmsemenov\OpenaiAttribute\OpenaiAttributeServiceProvider"
```
It is possible to specify request options in published config or in the model `generatedAttributes()` method.  
Available request options see https://platform.openai.com/docs/api-reference/chat/create
``` php
    // OpenAI api key
    'api_key' => env('OPENAI_API_KEY'),

    // Queue name for generate attributes jobs
    'queue_name' => env('OPENAI_QUEUE_NMAE', 'openai_generate'),

    // Default options for api chat completions requests.
    'default_options' => [
        'model' => env('OPENAI_API_MODEL', 'gpt-3.5-turbo'),
        'temperature' => 0.3,
        'max_tokens' => 100,
        'top_p' => 1.0,
        'frequency_penalty' => 0.0,
        'presence_penalty' => 0.0
    ],
```

## Usage
Trait `HasGeneratedAttributes` should be added to model. it contains list of model attributes with prompt and options (if needed, see default_options):
``` php
public function generatedAttributes(): array
{
    return [
        'description' => [
            'prompt' => 'Tell info about [model:name]". Wrap each paragraph to tag <p>',
            'max_tokens' => 1000
        ],

        'slug' => [
            'prompt' => 'Generate slug from [model:name]',
            'max_tokens' => 10
        ],
    ];
}
```

Models that fits condition in `[MyModel]->NeedsGenerateAttributes()` will be selected for attributes generation. Default: all.

Artisan commands to generate attributes:
``` bash
php artisan openai:generate App\\Models\\[MyModel] --queued
```
or generate instantly without queue option:
``` bash
php artisan openai:generate App\\Models\\[MyModel]
```
or specify model ids list:
``` bash
php artisan openai:generate App\\Models\\[MyModel] --ids=1 --ids=2
```
Other useful commands:
``` bash
Display queue size:
php artisan queue:monitor openai_generate

Remove jobs from queue:
php artisan queue:clear --queue=openai_generate

Run workers:
php artisan queue:work --queue=openai_generate
```

## Change log

Please see the [changelog](changelog.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [contributing.md](contributing.md) for details and a todolist.

## Security

If you discover any security related issues, please email dimitr.semenov@gmail.com instead of using the issue tracker.

## Credits

- [Dmitry Semenov][link-author]
- [All Contributors][link-contributors]

## License

MIT. Please see the [license file](license.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/dmsemenov/openai-attribute.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/dmsemenov/openai-attribute.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/dmsemenov/openai-attribute/master.svg?style=flat-square
[ico-styleci]: https://styleci.io/repos/12345678/shield

[link-packagist]: https://packagist.org/packages/dmsemenov/openai-attribute
[link-downloads]: https://packagist.org/packages/dmsemenov/openai-attribute
[link-travis]: https://travis-ci.org/dmsemenov/openai-attribute
[link-styleci]: https://styleci.io/repos/12345678
[link-author]: https://github.com/dmsemenov
[link-contributors]: ../../contributors


changelog.md
# Changelog

All notable changes to `OpenaiAttribute` will be documented in this file.

## Version 1.0

### Added
- Everything


src/OpenaiAttribute.php
<?php

namespace Dmsemenov\OpenaiAttribute;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Arr;
use Illuminate\Support\Facades\Log;
use Illuminate\Support\Str;
use OpenAI;

class OpenaiAttribute
{
    /**
     * Generate prompt from model
     *
     * @param Model $model
     * @return ?string
     */
    public function getPrompt(Model $model, string $attributeName): ?string
    {
        try {
            $prompt = Arr::get($model->generatedAttributes(), "{$attributeName}.prompt");

            $replaced = preg_replace_callback(
                "/\[model:\w*\]/",
                function ($matches) use ($model) {
                    return $model->{Str::between(current($matches), 'model:', ']')};
                },
                $prompt
            );
        } catch (\Throwable $e) {
            Log::error($e->getMessage());
            return null;
        }

        return $replaced;
    }

    /**
     * Fill generated attributes for model.
     *
     * @param Model $model
     * @return void
     */
    public function generate(Model $model): void
    {
        $modelUpdated = false;

        foreach ($model->generatedAttributes() as $attributeName => $attributeConfig) {
            $prompt = $this->getPrompt($model, $attributeName);
            if (empty($prompt)) {
                continue;
            }

            $requestBody = array_merge(config('openai-attribute.default_options'), Arr::except($attributeConfig, ['prompt']));

            $client = OpenAI::client(config('openai-attribute.api_key'));
            try {
                $result = $client->chat()->create([
                    ...$requestBody,
                    'messages' => [
                        ['role' => 'user', 'content' => $prompt]
                    ]
                ]);
            } catch (\Throwable $exception) {
                Log::error($exception->getMessage());
                return;
            }

            if ($text = Arr::get($result, 'choices.0.message.content')) {
                $modelUpdated = true;
                $model->generatedTextAlter($attributeName, $text);
            }
        }

        if ($modelUpdated) {
            $model->save();
        }
    }
}


src/Jobs/GenerateOpenAiAttributes.php
<?php

namespace Dmsemenov\OpenaiAttribute\Jobs;

use Dmsemenov\OpenaiAttribute\Facades\OpenaiAttribute;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\Middleware\WithoutOverlapping;
use Illuminate\Queue\SerializesModels;

class GenerateOpenAiAttributes implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    /**
     * Create a new job instance.
     */
    public function __construct(
        public Model $model
    ) {
        $this->onQueue(config('openai-attribute.queue_name'));
    }

    /**
     * Execute the job.
     */
    public function handle(): void
    {
        OpenaiAttribute::generate($this->model);
    }

    /**
     * Get the middleware the job should pass through.
     *
     * @return array<int, object>
     */
    public function middleware(): array
    {
        return [new WithoutOverlapping($this->model->id)];
    }
}


src/OpenaiAttributeServiceProvider.php
<?php

namespace Dmsemenov\OpenaiAttribute;

use Dmsemenov\OpenaiAttribute\Console\Commands\OpenAiGenerate;
use Illuminate\Support\ServiceProvider;

class OpenaiAttributeServiceProvider extends ServiceProvider
{
    /**
     * Perform post-registration booting of services.
     *
     * @return void
     */
    public function boot(): void
    {
        // $this->loadTranslationsFrom(__DIR__.'/../resources/lang', 'dmsemenov');
        // $this->loadViewsFrom(__DIR__.'/../resources/views', 'dmsemenov');
        // $this->loadMigrationsFrom(__DIR__.'/../database/migrations');
        // $this->loadRoutesFrom(__DIR__.'/routes.php');

        // Publishing is only necessary when using the CLI.
        if ($this->app->runningInConsole()) {
            $this->bootForConsole();
        }
    }

    /**
     * Register any package services.
     *
     * @return void
     */
    public function register(): void
    {
        $this->mergeConfigFrom(__DIR__.'/../config/openai-attribute.php', 'openai-attribute');

        // Register the service the package provides.
        $this->app->singleton('openai-attribute', function ($app) {
            return new OpenaiAttribute;
        });
    }

    /**
     * Get the services provided by the provider.
     *
     * @return array
     */
    public function provides()
    {
        return ['openai-attribute'];
    }

    /**
     * Console-specific booting.
     *
     * @return void
     */
    protected function bootForConsole(): void
    {
        // Publishing the configuration file.
        $this->publishes([
            __DIR__.'/../config/openai-attribute.php' => config_path('openai-attribute.php'),
        ], 'openai-attribute.config');

        // Publishing the views.
        /*$this->publishes([
            __DIR__.'/../resources/views' => base_path('resources/views/vendor/dmsemenov'),
        ], 'openai-attribute.views');*/

        // Publishing assets.
        /*$this->publishes([
            __DIR__.'/../resources/assets' => public_path('vendor/dmsemenov'),
        ], 'openai-attribute.views');*/

        // Publishing the translation files.
        /*$this->publishes([
            __DIR__.'/../resources/lang' => resource_path('lang/vendor/dmsemenov'),
        ], 'openai-attribute.views');*/

        // Registering package commands.
        $this->commands([
            OpenAiGenerate::class
        ]);
    }
}


src/Console/Commands/OpenAiGenerate.php
<?php

namespace Dmsemenov\OpenaiAttribute\Console\Commands;

use Dmsemenov\OpenaiAttribute\Jobs\GenerateOpenAiAttributes;
use Dmsemenov\OpenaiAttribute\Facades\OpenaiAttribute;
use Illuminate\Console\Command;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;

class OpenAiGenerate extends Command
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'openai:generate
                            {model : Eloquent model like App/User}
                            {--ids=* : Eloquent model id}
                            {--queued}';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Fill existing model attributes with OpenAi answer';

    /**
     * Model instance
     *
     * @var Model
     */
    protected Model $model;

    /**
     * Flag process as queue jobs
     *
     * @var bool
     */
    protected bool $queued;

    /**
     * Model ids to process
     *
     * @var array
     */
    protected array $ids;

    /**
     * Execute the console command.
     */
    public function handle()
    {
        $model = $this->argument('model');

        $this->queued = $this->option('queued');
        $this->ids = $this->option('ids');

        try {
            $this->model = new $model;
        } catch (\Throwable $e) {
            $this->error('Cant\'t resolve model: ' . $model);
            exit(1);
        }

        $query = $this->model->newModelQuery();

        if (!empty($this->ids)) {
            $query = $query->whereIn($this->model->getKeyName(), $this->ids);
        }

        if ($this->queued) {
            $this->handleQueued($query);
            return;
        }

        $this->handleInstantly($query);
    }

    /**
     * Handle request with queue jobs
     *
     * @param Builder $query
     * @return void
     */
    protected function handleQueued(Builder $query)
    {
        $this->info('OpenAi attribute generation added to queue.');
        $models = $query->get()->filter(fn($model) => $model->NeedsGenerateAttributes());

        foreach ($models as $model) {
            GenerateOpenAiAttributes::dispatch($model);
        }
    }


    /**
     * Handle request instantly
     *
     * @param Builder $query
     * @return void
     */
    protected function handleInstantly(Builder $query)
    {
        $this->info('OpenAi attribute generation started instantly.');

        $models = $query->get()->filter(fn($model) => $model->NeedsGenerateAttributes());

        foreach ($models as $model) {
            OpenaiAttribute::generate($model);
        }
    }
}


src/HasGeneratedAttributes.php
<?php

namespace Dmsemenov\OpenaiAttribute;

trait HasGeneratedAttributes
{
    /**
     * List of model generated attributes with prompt
     * Example:
     * 'description' => [
     *      'prompt' => 'Say something about [model:title]',
     *      'temperature' => 0.3,
     *      'max_tokens' => 1000,
     *      'top_p' => 1.0,
     *      'frequency_penalty' => 0.0,
     *      'presence_penalty' => 0.0
     * ]
     *
     * @return  array
     */
    public function generatedAttributes(): array
    {
        return [];
    }

    /**
     * Define condition if needed to generate model attributes presented in $generatedAttributes
     *
     * @return bool
     */
    public function NeedsGenerateAttributes(): bool
    {
        return true;
    }

    /**
     * Alter generated text if needed
     *
     * @param string $attributeName
     * @param string $text
     * @return void
     */
    public function generatedTextAlter(string $attributeName, string $text): void
    {
        $this->$attributeName = $text;
    }
}
